File: solutions/code-carl/ch07_backtracking/T4_LC0039_CombinationSum.java
/**
 * Practice template for:
 * - LeetCode 39. Combination Sum
 *   https://leetcode.cn/problems/combination-sum/
 */
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class T4_LC0039_CombinationSum {
    List<Integer> path = new ArrayList<>();
    List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        backtracking(candidates, target, 0, 0);
        return res;
    }

    public void backtracking(int[] nums, int target, int startIdx, int sum) {
        if (target == sum) {
            res.add(new ArrayList<>(path));
            return;
        }
        for (int i = startIdx; i < nums.length && sum + nums[i] <= target; i++) {
            path.add(nums[i]);
            backtracking(nums, target, i, sum + nums[i]);
            path.remove(path.size() - 1);
        }
    }
}

File: solutions/code-carl/ch07_backtracking/t4_lc0039_combination_sum.py
"""LeetCode 39. Combination Sum
https://leetcode.cn/problems/combination-sum/
"""

from typing import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        path: List[int] = []
        ans: List[List[int]] = []

        def backtracking(nums: List[int], target: int, sum: int, start_idx: int) -> None:
            if sum == target:
                ans.append(path[:])
                return

            for i in range(start_idx, len(nums)):
                if sum > target:
                    break
                path.append(nums[i])
                backtracking(nums, target, sum + nums[i], i)
                path.pop()

        if candidates:
            candidates.sort()
            backtracking(candidates, target, 0, 0)
        return ans


if __name__ == "__main__":
    solution = Solution()
    result = sorted(sorted(x) for x in solution.combinationSum([2, 3, 6, 7], 7))
    expected = sorted(sorted(x) for x in [[2, 2, 3], [7]])
    assert result == expected

    solution = Solution()
    result = sorted(sorted(x) for x in solution.combinationSum([2, 3, 5], 8))
    expected = sorted(sorted(x) for x in [[2, 2, 2, 2], [2, 3, 3], [3, 5]])
    assert result == expected

    solution = Solution()
    assert solution.combinationSum([2], 1) == []

    print("All tests passed.")
